text/0000-javascript-module-api.md

```diff
@@ -0,0 +1,1058 @@
+[migrator](#migration) on them, then replace the extracted code with the updated
+version. For the API docs, this tool would need to be able to handle Markdown embedded
+in JSDoc-style documentation.
+
+The benefit of this approach is that, once we have verified the script works
+reliably, we can wait until the last possible moment to make the switch. If we
+attempt to update everything by hand, the duration and tediousness of that
+process will likely take out an effective "lock" on the documentation code base,
+where people will put off making big changes because of the potential for merge
+conflicts.
+
+## API Documentation
+
+Our API documentation has long been a source of frustration, because the laundry
+list of (often rarely used or internal) classes makes Ember feel far more
+overwhelming than it really is.
+
+The shift to modules gives us a good opportunity to rethink the presentation of
+our API documentation. Instead of the imposing mono-list, we should group the
+API documentation by package–which, conveniently in this proposal, means they
+will also be grouped by area of functionality.
+
+We should investigate the broader ecosystem to see if there is a good tool that
+generates package-oriented documentation for JavaScript projects. If not, we may
+wish to adapt an existing tool to do so.
+
+## Explaining the Migration
+
+Once the guides and API documentation are updated, modules should be
+straightforward for new learners—indeed, more and more new learners are starting
+with JavaScript modules as the baseline.
+
+The most challenging aspect of teaching the new modules API, counterintuitively,
+will likely be _existing_ users. In particular, for changes that touch nearly
+every file, most teams working on large apps cannot pause work for a week to
+implement the change.
+
+Our focus needs to be:
+
+* Communicating clearly that the existing global build will work for the
+  foreseeable future.
+* Making clear the file size benefits of moving to modules.
+* Building robust tooling that allows even large apps to migrate in a day or
+  two, not a week.
+
+It is important to frame the module transition as a carrot, not a stick. We
+should avoid dire warnings or deprecation notices. Instead, we should provide
+good reporting when doing Ember CLI builds. If an app is compiled in globals
+mode, we can offer suggestions for how to reduce the file size, providing a
+helpful pointer to the modules migration guide. This will make the transition
+feel less like churn and more like an optimization opportunity that developers
+can take advantage of when they have the time or resources.
+
+### Addons
+
+One pitfall is that a _single_ use of the `Ember` global means we have to
+include the entire framework. That means that a developer could migrate their
+entire app to modules, but a single old addon that uses the Ember globals will
+negate the benefits.
+
+This requires a two-pronged strategy:
+
+* Tight integration into Ember CLI
```

Comment:
Offering a build-time transform is, I think, a brilliant solution to the addon compatibility question I hadn't considered, though it seems obvious now in retrospect.

I don't think it's a good idea for apps, because the process of going from globals to modules is pretty expensive, especially if we do a very thorough job of having the codemod handle the edge cases people discover.

However, the reverse is not true: converting from modules back to globals is trivially easy. If an addon updates to modules, we can provide a Broccoli transform that will rewrite them to globals if the app is on an older version of Ember. I think we can make this step pretty fast, and of course it is eminently cacheable so subsequent builds shouldn't have any performance impact at all.

As for apps, I think we will have an addon that imports everything and constructs the old `Ember` global for you. If we detect that you haven't `import`ed anything, we can add this an implicit import, with an option to explicitly opt-in if that detection doesn't work for some reason.